rfd/headless-login.md

```diff
@@ -0,0 +1,367 @@
+
+Since the user certificates are only meant to last for a single client lifetime, the Auth server should issue the certificates with a 1 minute TTL. This will prevent faulty clients or bugs from being exploited to hold headless authenticated certificates beyond the lifetime of a normal client.
+
+3. **User must complete WebAuthn verification for each individual `tsh` request (`tsh ls`, `tsh ssh`, etc.).**
+
+Headless authentication, like any login mechanism, can be started by any unauthenticated user. To prevent phishing attacks, we must prompt the user to acknowledge and approve each headless authentication request with WebAuthn, since WebAuthn provides strong protection against phishing attacks. Legacy OTP MFA methods will not be supported.
+
+4. **The Server must verify that the client requesting certificates is the only client that can use the certificates.**
+
+Depending on the headless authentication API flow, it may be split into multiple requests. It is crucial that the server verifies that the client who requested headless authentication is the only client that can use the resulting credentials. This will be accomplished with a PKCE-like flow, where the client provides their public key in the initial request, and the server ultimately issues certificates for the original client public key. Additionally, the user will be asked to sign for the public key before the server issues the certificates (TODO: is this possible without x509 certificate + TLS handshake)
+
+Note: Ideally, we would ensure that the only client that can *receive* certificates is the requesting client, but ensuring the PKCE flow above is enough to defend against attacks since the certificates are useless without the client's private key. If we want to remove the possibility of an attack attempting to intercept the certificates, we can use an HTTP websocket or gRPC stream endpoint to encapsulate the headless authentication request into a single API request. This would require some additional implementation details to be flushed out, so it is not in this initial draft design.
+
+### Headless authentication overview
+
+The headless authentication flow is shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local Machine
+    participant Remote Machine
+    participant Teleport Proxy
+    participant Teleport Auth
+
+    Note over Remote Machine: tsh --headless ssh user@node01
+    Remote Machine->>+Teleport Proxy: initiate headless login
+    Teleport Proxy->>Teleport Auth: save request with new uuid
+    Note over Teleport Auth: Request: {id, user, publicKey, ...}
+    Teleport Auth->>Teleport Proxy: request ID
+    Teleport Proxy->>-Remote Machine: request ID
+    Note over Remote Machine: print command + URL
+
+    Remote Machine->>+Teleport Proxy: request signed user certs<br/>for request ID
+    Note over Local Machine: user copies command / URL<br/>to local terminal / browser
```

Comment:
I think this point is missing that the user must be logged in locally to use the link.

Reply:
Agree, this does not show that users have to be logged in with their browser or cli to use the link